x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/search/BasicQueryClient.java

```diff
@@ -64,48 +68,52 @@ protected void search(SearchRequest search, ActionListener<SearchResponse> liste
     }
 
     @Override
-    public void get(Iterable<List<HitReference>> refs, ActionListener<List<List<GetResponse>>> listener) {
-        MultiGetRequestBuilder requestBuilder = client.prepareMultiGet();
-        // no need for real-time
-        requestBuilder.setRealtime(false)
-                      .setRefresh(false);
+    public void fetchHits(Iterable<List<HitReference>> refs, ActionListener<List<List<SearchHit>>> listener) {
+        IdsQueryBuilder idsQuery = idsQuery();
 
-        int sz = 0;
+        int innerListSize = 0;
+        Set<String> indices = new HashSet<>();
+
+        // associate each reference with its own
+        final Map<HitReference, List<Integer>> referenceToPosition = new HashMap<>();
+        int counter = 0;
 
         for (List<HitReference> list : refs) {
-            sz = list.size();
+            innerListSize = list.size();
             for (HitReference ref : list) {
-                Item item = new Item(ref.index(), ref.id());
-                // make sure to get the whole source
-                item.fetchSourceContext(FetchSourceContext.FETCH_SOURCE);
-                requestBuilder.add(item);
+                idsQuery.addIds(ref.id());
```

Comment:
Is the ref.id() unique across indices? Is there a possibility to fetch more than one document residing in different indices?

Reply:
Of course not. You can have a document with id 1 - that's not unique across indices hence why the results are checked based on their id and index.

Reply:
As discussed, I'm worried about some edge case where we have multiple indices in the result docs and multiple id collisions which could lead to fetching many more docs than required, which will also increase the memory consumption. On the other hand, I don't know that the performance impact would be if we make multiple IdQuery requests, each for each index involved along with its corresponding doc ids.